src/main/cpp/src/datetime_parser.cu

```diff
@@ -0,0 +1,672 @@
+
+    return {sign * (hms[0] * 3600L + hms[1] * 60L + hms[2]), ParseUtcLikeTzResult::UTC_LIKE_TZ};
+  }
+
+  /**
+   * use binary search to find tz index.
+   */
+  __device__ inline int query_index_from_tz_db(string_view const& tz_lit) const
+  {
+    auto const tz_col                  = tz_indices->child(0);
+    auto const index_in_transition_col = tz_indices->child(1);
+
+    auto string_iter_begin =
+      thrust::make_transform_iterator(thrust::make_counting_iterator(0), get_string_fn{tz_col});
+    auto string_iter_end = string_iter_begin + tz_col.size();
+    auto it              = thrust::lower_bound(
+      thrust::seq, string_iter_begin, string_iter_end, tz_lit, thrust::less<string_view>());
+    if (it != string_iter_end && *it == tz_lit) {
+      // found tz
+      auto tz_name_index = static_cast<size_type>(it - string_iter_begin);
+      return static_cast<int>(index_in_transition_col.element<int32_t>(tz_name_index));
+    } else {
+      // not found tz
+      return -1;
+    }
+  }
+
+  /**
+   * Perform binary search to search out the offset from UTC based on local epoch
+   * instants. Basically, this is the same approach as
+   * `convert_timestamp_tz_functor`.
+   */
+  __device__ inline int64_t compute_utc_offset(int64_t loose_epoch_second, size_type tz_index) const
+  {
+    auto const& utc_offsets    = transitions->child().child(2);
+    auto const& loose_instants = transitions->child().child(3);
+
+    auto const local_transitions = cudf::list_device_view{*transitions, tz_index};
+    auto const list_size         = local_transitions.size();
+
+    auto const transition_times = cudf::device_span<int64_t const>(
+      loose_instants.data<int64_t>() + local_transitions.element_offset(0),
+      static_cast<size_t>(list_size));
+
+    auto const it = thrust::upper_bound(
+      thrust::seq, transition_times.begin(), transition_times.end(), loose_epoch_second);
+    auto const idx         = static_cast<size_type>(thrust::distance(transition_times.begin(), it));
+    auto const list_offset = local_transitions.element_offset(idx - 1);
+    return static_cast<int64_t>(utc_offsets.element<int32_t>(list_offset));
+  }
+
+  /**
+   * The formula to compute loose epoch from local time. The loose epoch is used
+   * to search for the corresponding timezone offset of specific zone ID from
+   * TimezoneDB. The target of loose epoch is to transfer local time to a number
+   * which is proportional to the real timestamp as easily as possible. Loose
+   * epoch, as a computation approach, helps us to align probe(kernel side) to
+   * the TimezoneDB(Java side). Then, we can apply binary search based on loose
+   * epoch instants of TimezoneDB to find out the correct timezone offset.
+   */
+  __device__ inline int64_t compute_loose_epoch_s(timestamp_components const& ts) const
+  {
+    return (ts.year * 400 + (ts.month - 1) * 31 + ts.day - 1) * 86400L + ts.hour * 3600L +
```

Comment:
lots of magic numbers in here. I can understand 60, 3600, 86400, and 31. Would be nice if they were commented or `constexpr`s. The 400 requires a bit of explanation.

Reply:
Added comments.